src/Raven.Server/SqlMigration/RavenDocumentFactory.cs

```diff
@@ -0,0 +1,193 @@
+﻿using System;
+using System.Collections.Generic;
+
+namespace Raven.Server.SqlMigration
+{
+    public class RavenDocumentFactory
+    {
+        public WriteOptions Options { get;}
+
+        public RavenDocumentFactory() : this(new WriteOptions()) {}
+
+        public RavenDocumentFactory(WriteOptions options)
+        {
+            Options = options;
+        }
+
+        public RavenDocument FromReader(SqlReader reader, SqlTable table, bool valuesOnly = false)
+        {
+            return FromReader(reader, table, out _, valuesOnly);
+        }
+
+        public RavenDocument FromReader(SqlReader reader, SqlTable table, out Dictionary<string, byte[]> attachments, bool valuesOnly = false)
+        {
+            attachments = new Dictionary<string, byte[]>();
+            var document = FromReaderInternal(reader, table, ref attachments, valuesOnly);
+            return document;
+        }
+
+        private RavenDocument FromReaderInternal(SqlReader reader, SqlTable table, ref Dictionary<string, byte[]> attachments, bool valuesOnly = false)
+        {
+            var document = new RavenDocument(table.Name);
+
+            var id = GetName(table.Name);
+
+            if (!table.IsEmbedded)
+                document.SetCollection(id);
+
+            for (var i = 0; i < reader.FieldCount; i++)
+            {
+                var columnName = reader.GetName(i);
+
+                object value;
+
+                try
+                {
+                    value = reader[i];
+                }
+                catch (Exception e)
+                {
+                    if (Options.SkipUnsopportedTypes)
+                        continue;
+
+                    throw new InvalidOperationException($"Cannot read column '{columnName}' in table '{table.Name}'. (Unsupported type: {reader.GetDataTypeName(i)})", e);
+                }
+
+                var isForeignKey = table.ForeignKeys.TryGetValue(columnName, out var foreignKeyTable);
+                var isNullOrEmpty = value is DBNull || string.IsNullOrWhiteSpace(value.ToString());
```

Comment:
You are forcing materializing of the value, which we probably don't need.
Instead, check if it is a string and avoid the call to `ToString`